expands platform lifecycle tutorial (#769)
buildpacks/docs: content/docs/for-platform-operators/tutorials/lifecycle/_index.md

include_summaries=true
+++

A `platform` orchestrates builds by invoking the [lifecycle][lifecycle] binary together with buildpacks and application source code to produce a runnable `OCI image`.

<!--more-->

The majority of Buildpack users use platforms, such as [pack][pack] and [kpack][kpack], to run Buildpacks and create `OCI images`. However this might not be desireable especially for users maintaining their own platforms and seeking more control over how the underlying Buildpack `lifecycle phases` are executed.

> In this step-by-step tutorial, you will build a `Bash` application without using any `platform` tooling like `pack` or `kpack`. You will also leverage the individual `lifecycle phases` to produce a runnable application image.

Reviewer: Can we explicitly state up-front that this guide derives from https://medium.com/buildpacks/unpacking-cloud-native-buildpacks-ff51b5a767bf?

Author: Sure. I had this note initially at the top; then removed it to the bottom because I didn't want to distract the reader. Added at the start.

## Prerequisites

You'll need to clone a local copy of the following to get started:

* The `lifecycle`: `git clone https://github.com/buildpacks/lifecycle`
* The official `Buildpack.io` samples repo: `git clone https://github.com/buildpacks/samples`

As previously mentioned, the `lifecycle` orchestrates `Buildpacks` then assembles the resulting artifacts into an `OCI image`.  The `lifecycle` is composed of a series of distinct `phases` that need to be executed to have the final image built and exported.

## Overview

Now that you’re set up, let’s build our `Bash` application and dive deeper into the `lifecycle` and its phases.

### Build the lifecycle

As a starting step, you need to build the `lifecycle` in order to use its phases. This could be done by navigating to the `lifecycle` directory and executing one of the following commands, depending on your system architecture.

* `make build` for `AMD64` architectures

> * `make build-darwin-arm64` for `ARM64` architectures

Reviewer: We may need to point out that `darwin-arm64` is specifically for Mac machines.  Do we also need to point out that the entire process is most easily followed on a Linux machine?

Author: added

### Examine lifecycle phases

A single app image build consists of the following phases:

1. [Analysis](#analyze)
2. [Detection](#detect)
3. [Cache Restoration](#restore)
4. [Build](#build)
5. [Export](#export)

> Note that a `platform` executes the phases above either by invoking phase-specific lifecycle binaries in order or by executing `/cnb/lifecycle/creator`.

Let's expand each `lifecycle` phase to explain how the `lifecycle` orchestrates buildpacks:

#### Analyze

The `analyze` phase runs before the `detect` phase in order to validate registry access for all images used during the `build` as early as possible. In this way it provides faster failures for end users.

Prior to executing `/cnb/lifecycle/analyzer`, you need to create two directories in the `root` directory as follows:

```text

mkdir -p apps/bash-script
mkdir -p layers

```

* `apps` directory that contains a `bash-script` directory
* `layers` directory that contains subdirectories representing each layer created by the Buildpack in the final image or build cache.

Next,  you need to copy the `bash-script` samples into our `apps/bash-script` directory, which will host our app's source code.

```text

> cp -r "$CNB_SAMPLES_PATH/apps/bash-script/" ./apps/bash-script

Reviewer: Let's use curlies to make the variables more explicit
```suggestion
cp -r "${CNB_SAMPLES_PATH}/apps/bash-script/" ./apps/bash-script
```

Author: updated

```

Now, you can invoke the `analyzer` for `AMD64` and `ARM64` architectures respectively

```text

> $CNB_LIFECYCLE_PATH/analyzer -log-level debug -daemon -layers="./layers" -run-image cnbs/sample-stack-run:bionic apps/bash-script

Reviewer: ```suggestion
${CNB_LIFECYCLE_PATH}/analyzer -log-level debug -daemon -layers="./layers" -run-image cnbs/sample-stack-run:bionic apps/bash-script
```

Author: updated

```

```text

> $CNB_LIFECYCLE_PATH/analyzer -log-level debug -daemon -layers="./layers" -run-image arm64v8/ubuntu:latest apps/bash-script

Reviewer: ```suggestion
${CNB_LIFECYCLE_PATH}/analyzer -log-level debug -daemon -layers="./layers" -run-image arm64v8/ubuntu:latest apps/bash-script
```

Note to @AidanDelaney, double check this run image.

Author: updated

  ```

* The official `Buildpack.io` samples repo
  
  ```text
  git clone https://github.com/buildpacks/samples
  ```

As previously mentioned, the `lifecycle` orchestrates `Buildpacks` then assembles the resulting artifacts into an `OCI image`.  The `lifecycle` is composed of a series of distinct `phases` that need to be executed to have the final image built and exported.

## Overview

Now that you’re set up, let’s build our `Bash` application and dive deeper into the `lifecycle` and its phases.

### Build the lifecycle

As a starting step, you need to build the `lifecycle` in order to use its phases. This could be done by navigating to the `lifecycle` directory and executing one of the following commands, depending on your system architecture.

* `make build` for `AMD64` architectures (for Linux users)
* `make build-darwin-arm64` for `ARM64` architectures (for Mac users)

> Please note that the entire process is most easily followed on Linux systems

### Set environment variables

In order to execute the various `lifecycle phases` correctly, you first need to set the values of few important environment variables by running the following commands in the terminal:

```text

> export CNB_USER_ID=1000 CNB_GROUP_ID=1000 CNB_PLATFORM_API=0.14

Author: updated

export CNB_SAMPLES_PATH="/<your-path>/samples"
export CNB_LIFECYCLE_PATH="/<your-path/lifecycle/out/<your-arch>/lifecycle"`

```

Where

> * `CNB_USER_ID` and `CNB_GROUP_ID` are arbitrary values that need to be consistent, which both have a default value of `1000`.

Author: updated

### Examine lifecycle phases

A single app image build consists of the following phases:

1. [Analysis](#analyze)
2. [Detection](#detect)
3. [Cache Restoration](#restore)
4. [Build](#build)
5. [Export](#export)

> Note that a `platform` executes the phases above either by invoking phase-specific lifecycle binaries in order or by executing `/cnb/lifecycle/creator`.

Let's expand each `lifecycle` phase to explain how the `lifecycle` orchestrates buildpacks:

#### Analyze

The `analyze` phase runs before the `detect` phase in order to validate registry access for all images used during the `build` as early as possible. In this way it provides faster failures for end users.

Prior to executing `/cnb/lifecycle/analyzer`, you need to create two directories in the `root` directory as follows:

```text
mkdir -p apps/bash-script
mkdir -p layers
```

* `apps` directory that contains a `bash-script` directory
* `layers` directory that contains subdirectories representing each layer created by the Buildpack in the final image or build cache.

Next,  you need to copy the `bash-script` samples into our `apps/bash-script` directory, which will host our app's source code.

```text

> cp -r "${CNB_SAMPLES_PATH}/apps/bash-script/" ./apps/bash-script

Author: updated

```

Now, you can invoke the `analyzer` for `AMD64` and `ARM64` architectures respectively

```text

> ${CNB_LIFECYCLE_PATH}/analyzer -log-level debug -daemon -layers="./layers" -run-image cnbs/sample-stack-run:bionic apps/bash-script

Author: updated

```

```text
${CNB_LIFECYCLE_PATH}/analyzer -log-level debug -daemon -layers="./layers" -run-image arm64v8/ubuntu:latest apps/bash-script
```

The commands above run the `analyzer` with:

* A `debug` logging level
* Pointing to the local `Docker daemon`
* Pointing to the `layers` directory, which is the main `lifecycle` working directory
* Running the specified image
* The path to the app that you are analyzing

Now the `analyzer`:

* Checks a registry for previous images called `apps/bash-script`.
* Resolves the image metadata making it available to the subsequent `restore` phase.
* Verifies that you have write access to the registry to create or update the image called `apps/bash-script`.

In this tutorial, there is no previous `apps/bash-script` image, and the output produced should be similar to the following:

```text

> OUTPUT PLACEHOLDER

Author: added

```

> Now if you check the `layers` directory, you should have a `analyzer.toml` file with a few null entries.

Author: updated

#### Detect

> In this phase, the `detector` looks for an ordered group of buildpacks that will be used during the `build` phase. The `detector` requires an `order.toml` file being present in the `root` directory, which you could derived from `builder.toml` in the `samples` directory while removing the deprecated `stack` section as follows:

Author: updated

```text
cat "${CNB_SAMPLES_PATH}/builders/jammy/builder.toml" | grep -v -i "stack" | sed 's/\.\.\/\.\./\./' > order.toml

```

`order.toml` files contain a list of groups with each group containing a list of buildpacks. The `detector` reads `order.toml` and looks for the first group that passes the detection process.

##### Set buildpacks layout directory

Before running the `detector`, you need to:

1. Create a `buildpacks` directory in the `root` directory

    ```text
    mkdir -p buildpacks
    ```

2. Then you must populate the `buildpacks` directory with your buildpacks of interest.

    > You have to follow the [directory layout][directory layout] defined in the buildpack spec, where each top-level directory is a `buildpack ID` and each second-level directory is a `buildpack version`.

> 

Reviewer: ```suggestion

We will use [`dasel`](http://github.com/tomwright/dasel/) to help us parse toml files.

```command
$ go install github.com/tomwright/dasel/v2/cmd/dasel@master
```

Author: added

    Let’s do that for every buildpack in the `samples/buildpacks` directory:

    ```text
    for f in `ls --color=no $CNB_SAMPLES_PATH/buildpacks | grep -v README`
    do
    mkdir -p ./buildpacks/samples_"$f"/0.0.1
    cp -r "$CNB_SAMPLES_PATH/buildpacks/$f/" ./buildpacks/samples_"$f"/0.0.1/
    done

>     ```

Author: updated

```text
cat "${CNB_SAMPLES_PATH}/builders/jammy/builder.toml" | grep -v -i "stack" | sed 's/\.\.\/\.\./\./' > order.toml

```

`order.toml` files contain a list of groups with each group containing a list of buildpacks. The `detector` reads `order.toml` and looks for the first group that passes the detection process.

##### Set buildpacks layout directory

Before running the `detector`, you need to:

1. Create a `buildpacks` directory in the `root` directory

    ```text
    mkdir -p buildpacks
    ```

2. Then you must populate the `buildpacks` directory with your buildpacks of interest.

    > You have to follow the [directory layout][directory layout] defined in the buildpack spec, where each top-level directory is a `buildpack ID` and each second-level directory is a `buildpack version`.

    We will use [`dasel`](http://github.com/tomwright/dasel/) to help us parse toml files.

    ```command
    $ go install github.com/tomwright/dasel/v2/cmd/dasel@master

    Let’s do that for every buildpack in the `samples/buildpacks` directory:

    ```text
    for f in $(ls --color=no ${CNB_SAMPLES_PATH}/buildpacks | grep -v README)
    do

>     bp_version=$(cat $f | dasel -r toml "${CNB_SAMPLES_PATH}/buildpacks/${f}/buildpack.version" | sed s/\'//g)

Author: updated

    mkdir -p ./buildpacks/samples_"${f}"/${bp_version}

>     cp -r "$CNB_SAMPLES_PATH/buildpacks/${f}/" ./buildpacks/samples_"${f}"/${bp_version}/

Reviewer: ```suggestion
    cp -r "$CNB_SAMPLES_PATH/buildpacks/${f}/*" ./buildpacks/samples_"${f}"/${bp_version}/
```

Author: updated

    done
    ```

Now, you can run the `detector` binary:

```text
${CNB_LIFECYCLE_PATH}/detector -log-level debug -layers="./layers" -order="./order.toml" -buildpacks="./buildpacks" -app apps/bash-script
```

> The output of the above command should have `group.toml` and `plan.toml` output files (i.e., the groups that have passed the detection have been written into the `group.toml` file writing its build plan into the `plan.toml` file)

Reviewer: ```suggestion
The output of the above command should have `layers/group.toml` and `layers/plan.toml` output files (i.e., the groups that have passed the detection have been written into the `group.toml` file writing its build plan into the `plan.toml` file)
```

Author: updated

```text

> OUTPUT PLACEHOLDER

Reviewer: ```suggestion
Starting detector...
Parsing inputs...
Ensuring privileges...
Executing command...
Timer: Detector started at 2024-10-01T07:00:50Z
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
======== Results ========
fail: samples/java-maven@0.0.2
======== Results ========
fail: samples/kotlin-gradle@0.0.2
======== Results ========
fail: samples/ruby-bundler@0.0.1
======== Results ========
pass: samples/hello-world@0.0.1
pass: samples/hello-moon@0.0.1
Resolving plan... (try #1)
samples/hello-world 0.0.1
samples/hello-moon  0.0.1
Timer: Detector ran for 26.011769ms and ended at 2024-10-01T07:00:50Z
```

Author: added

```

You can view more details about the [order](https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/detect/#ordertoml), [group](https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/detect/#grouptoml) and [plan](https://buildpacks.io/docs/concepts/components/lifecycle/detect/#plantoml) toml files in the platform documentation.

#### Restore

The `restorer` retrieves cache contents, if it contains any, into the build container. During this phase, the `restorer` looks for layers that could be reused or should be replaced while building the application image.

First, you need to create the `cache` directory, and then run the `restorer` binary as added below:

```text
mkdir cache
```

```text
${CNB_LIFECYCLE_PATH}/restorer -log-level debug -layers="./layers" -group="./layers/group.toml" -cache-dir="./cache" -analyzed="./layers/analyzed.toml"
```

The `cache` directory should now be populated by two sub-directories, `committed` and `staging` as shown in the output below:

```text

> OUTPUT PLACEHOLDER

Reviewer: ```suggestion
Starting restorer...
Parsing inputs...
Ensuring privileges...
Executing command...
No run metadata found at path "/cnb/run.toml"
Run image info in analyzed metadata is: 
{"Reference":"","Image":"cnbs/sample-stack-run:jammy","Extend":false,"target":{"os":"linux","arch":"amd64"}}
Timer: Restorer started at 2024-10-01T07:03:47Z
Restoring Layer Metadata
Reading buildpack directory: /tmp/example/layers/samples_hello-world
Reading buildpack directory: /tmp/example/layers/samples_hello-moon
Reading Buildpack Layers directory /tmp/example/layers
Reading buildpack directory: /tmp/example/layers/samples_hello-world
Reading Buildpack Layers directory /tmp/example/layers
Reading buildpack directory: /tmp/example/layers/samples_hello-moon
Timer: Restorer ran for 274.41µs and ended at 2024-10-01T07:03:47Z
```

Author: added

```

#### Build

The `builder` transforms application source code into runnable artifacts that can be packaged into a container.

Before running the `builder`, the following steps are required:

1. Create two directories:
   * `platform` directory to store configurations and environment variables
   * `workspace` directory to store application source code and where you build it
  
    ```text
    mkdir -p platform
    mkdir -p workspace
    ```

2. Copy the source code from the `app` directory to the `workspace` directory

    ```text
    cp -r apps/bash-script/* ./workspace
    ```

3. Create a `launcher` file with instructions to run your application

>     ```text

Reviewer: ```suggestion
    ```text
    mkdir -p layers/samples_hello-moon
```

Author: updated

    cat << EOF > ./layers/samples_hello-moon/launch.toml
    [[processes]]
    type = "shell"
    command = ["./app.sh"]
    EOF
    ```

Now you're ready to run the `builder` as follows:

```text
${CNB_LIFECYCLE_PATH}/builder -log-level debug -layers="./layers" -group="./layers/group.toml" -analyzed="./layers/analyzed.toml" -plan="./layers/plan.toml" -buildpacks="./buildpacks" -app="./workspace" -platform="./platform"
```

Taking a deep look at the following output shows that you have built the two buildpacks that we need in order to run our `bash-script` application. In addition, checking the `layers` directory should show other directories like the two from the buildpacks, a `config` and a `sbom` ones.

```text

> OUTPUT PLACEHOLDER

Reviewer: ```suggestion
Starting builder...
Parsing inputs...
Ensuring privileges...
Executing command...
Timer: Builder started at 2024-10-01T07:07:46Z
target distro name/version labels not found, reading /etc/os-release file
Running build for buildpack samples/hello-world@0.0.1
Looking up buildpack
Finding plan
Creating plan directory
Preparing paths
Running build command
---> Hello World buildpack
     platform_dir files:
       /tmp/example/platform:
       total 0
       drwxr-xr-x 2 gitpod gitpod  40 Oct  1 07:04 .
       drwxr-xr-x 8 gitpod gitpod 180 Oct  1 07:04 ..
     env_dir: /tmp/example/platform/env
     env vars:
       declare -x CNB_BP_PLAN_PATH="/tmp/samples_hello-world-576309032/samples_hello-world/plan.toml"
       declare -x CNB_BUILDPACK_DIR="/tmp/example/buildpacks/samples_hello-world/0.0.1"
       declare -x CNB_LAYERS_DIR="/tmp/example/layers/samples_hello-world"
       declare -x CNB_PLATFORM_DIR="/tmp/example/platform"
       declare -x CNB_TARGET_ARCH="amd64"
       declare -x CNB_TARGET_DISTRO_NAME="ubuntu"
       declare -x CNB_TARGET_DISTRO_VERSION="22.04"
       declare -x CNB_TARGET_OS="linux"
       declare -x HOME="/home/gitpod"
       declare -x HOSTNAME="buildpacks-docs-dusxugo5ehi"
       declare -x OLDPWD
       declare -x PATH="/bin:/usr/bin:/usr/local/bin"
       declare -x PWD="/tmp/example/workspace"
       declare -x SHLVL="1"
     layers_dir: /tmp/example/layers/samples_hello-world
     plan_path: /tmp/samples_hello-world-576309032/samples_hello-world/plan.toml
     plan contents:
       [[entries]]
         name = "some-world"
       
       [[entries]]
         name = "some-world"
         [entries.metadata]
           world = "Earth-616"

---> Done
Processing layers
Updating environment
Reading output files
Updating buildpack processes
Updating process list
Finished running build for buildpack samples/hello-world@0.0.1
Running build for buildpack samples/hello-moon@0.0.1
Looking up buildpack
Finding plan
Creating plan directory
Preparing paths
Running build command
---> Hello Moon buildpack
     env_dir: /tmp/example/platform/env
     env vars:
       declare -x CNB_BP_PLAN_PATH="/tmp/samples_hello-moon-3356528850/samples_hello-moon/plan.toml"
       declare -x CNB_BUILDPACK_DIR="/tmp/example/buildpacks/samples_hello-moon/0.0.1"
       declare -x CNB_LAYERS_DIR="/tmp/example/layers/samples_hello-moon"
       declare -x CNB_PLATFORM_DIR="/tmp/example/platform"
       declare -x CNB_TARGET_ARCH="amd64"
       declare -x CNB_TARGET_DISTRO_NAME="ubuntu"
       declare -x CNB_TARGET_DISTRO_VERSION="22.04"
       declare -x CNB_TARGET_OS="linux"
       declare -x HOME="/home/gitpod"
       declare -x HOSTNAME="buildpacks-docs-dusxugo5ehi"
       declare -x OLDPWD
       declare -x PATH="/bin:/usr/bin:/usr/local/bin"
       declare -x PWD="/tmp/example/workspace"
       declare -x SHLVL="1"
     layers_dir: /tmp/example/layers/samples_hello-moon
     plan_path: /tmp/samples_hello-moon-3356528850/samples_hello-moon/plan.toml
     plan contents:

---> Done
Processing layers
Updating environment
Reading output files
Updating buildpack processes
Updating process list
Finished running build for buildpack samples/hello-moon@0.0.1
Copying SBOM files
Creating SBOM files for legacy BOM
Listing processes
Timer: Builder ran for 20.200892ms and ended at 2024-10-01T07:07:46Z
```

Author: added

```

#### Export

The purpose of the `export` phase is to create a new `OCI` image using a combination of remote layers, local `<layers>/<layer>` layers, and the processed `app` directory.

To export the artifacts built by the `builder`, you first need to specify the path of the `launcher` that your image is going to run:

* For AMD64 architectures
  
  ```text

>   export {CNB_LINUX_LAUNCHER_PATH}=/<your-path>/lifecycle/out/linux-amd64/lifecycle/launcher

Reviewer: ```suggestion
  export CNB_LINUX_LAUNCHER_PATH=/<your-path>/lifecycle/out/linux-amd64/lifecycle/launcher
```

Author: updated

  ```

* For ARM64 Architectures

  ```text

>   export {CNB_LINUX_LAUNCHER_PATH}=/<your-path>/lifecycle/out/linux-arm64/lifecycle/launcher

Reviewer: ```suggestion
  export CNB_LINUX_LAUNCHER_PATH=/<your-path>/lifecycle/out/linux-arm64/lifecycle/launcher
```

Author: updated

include_summaries=true
+++

A `platform` orchestrates builds by invoking the [lifecycle][lifecycle] binary together with buildpacks and application source code to produce a runnable `OCI image`.

<!--more-->

The majority of Buildpack users use platforms, such as [pack][pack] and [kpack][kpack], to run Buildpacks and create `OCI images`. However this might not be desireable especially for users maintaining their own platforms and seeking more control over how the underlying Buildpack `lifecycle phases` are executed.

> This tutorial is derived from a [blog post][blog post] contributed by one of our community.

In this step-by-step tutorial, you will build a `Bash` application without using any `platform` tooling like `pack` or `kpack`. You will also leverage the individual `lifecycle phases` to produce a runnable application image.

## Prerequisites

> You'll need to clone a local copy of the following to get started:

Reviewer: Should we note the platforms on which we know this tutorial works?

```suggestion
This tutorial has been tested on linux amd64 and darwin arm64 platforms.  You'll need to clone a local copy of the following to get started:
```

* The `lifecycle`
  
  ```text
  git clone https://github.com/buildpacks/lifecycle
  ```

* The official `Buildpack.io` samples repo
  
  ```text
  git clone https://github.com/buildpacks/samples
  ```

As previously mentioned, the `lifecycle` orchestrates `Buildpacks` then assembles the resulting artifacts into an `OCI image`.  The `lifecycle` is composed of a series of distinct `phases` that need to be executed to have the final image built and exported.

## Overview

Now that you’re set up, let’s build our `Bash` application and dive deeper into the `lifecycle` and its phases.

### Build the lifecycle

As a starting step, you need to build the `lifecycle` in order to use its phases. This could be done by navigating to the `lifecycle` directory and executing one of the following commands, depending on your system architecture.

> * `make build` for `AMD64` architectures (for Linux users)

Reviewer: ```suggestion
* `make build-linux-amd64` for `AMD64` architectures (for Linux users)
```

* `make build-darwin-arm64` for `ARM64` architectures (for Mac users)

> > Please note that the entire process is most easily followed on Linux systems

Author: added

### Set environment variables

In order to execute the various `lifecycle phases` correctly, you first need to set the values of few important environment variables by running the following commands in the terminal:

```text

> export CNB_USER_ID=$(id -u) CNB_GROUP_ID=$(id -g) CNB_PLATFORM_API=0.14

Reviewer: The way this tutorial works, I think we're working on the host.  Therefore the `$(id -u)` should use the user id on the current host and allow the phases to create directories.

> export CNB_USER_ID=$(id -u) CNB_GROUP_ID=$(id -g) CNB_PLATFORM_API=0.14

Author: @AidanDelaney so there is no need to change anything here? thanks

> export CNB_USER_ID=$(id -u) CNB_GROUP_ID=$(id -g) CNB_PLATFORM_API=0.14

Reviewer: Nope, all good.

> export CNB_USER_ID=$(id -u) CNB_GROUP_ID=$(id -g) CNB_PLATFORM_API=0.14

Reviewer 2: I do think we should point out that we're running on a host machine and that is unusual for buildpacks but useful for teaching

> export CNB_USER_ID=$(id -u) CNB_GROUP_ID=$(id -g) CNB_PLATFORM_API=0.14

Author: added

* `CNB_PLATFORM_API` or the `Platform API` version, varies depending on the use case. This tutorial uses `v0.14`, which is the latest [Platform API][Platform API] version.
* `CNB_SAMPLES_PATH` represents the path of our local copy of the `samples` directory.
* `CNB_LIFECYCLE_PATH` represents the path of our local compiled `lifecycle` directory.

### Examine lifecycle phases

A single app image build consists of the following phases:

1. [Analysis](#analyze)
2. [Detection](#detect)
3. [Cache Restoration](#restore)
4. [Build](#build)
5. [Export](#export)

> Note that a `platform` executes the phases above either by invoking phase-specific lifecycle binaries in order or by executing `/cnb/lifecycle/creator`.

Let's expand each `lifecycle` phase to explain how the `lifecycle` orchestrates buildpacks:

#### Analyze

The `analyze` phase runs before the `detect` phase in order to validate registry access for all images used during the `build` as early as possible. In this way it provides faster failures for end users.

Prior to executing `/cnb/lifecycle/analyzer`, you need to create two directories in the `root` directory as follows:

```text
mkdir -p apps/bash-script
mkdir -p layers

> ```

Reviewer: I used `/tmp/tutorial` as the root for my run-through.

> ```

Reviewer: Is an simple solution to create a tmp dir for all this work?

```
TUTORIAL_ROOT=$(mktemp -d)
cd ${TUTORIAL_ROOT}
...
```

> ```

Reviewer 2: Added code suggestions

* `apps` directory that contains a `bash-script` directory
* `layers` directory that contains subdirectories representing each layer created by the Buildpack in the final image or build cache.

Next,  you need to copy the `bash-script` samples into our `apps/bash-script` directory, which will host our app's source code.

```text

> cp -r "${CNB_SAMPLES_PATH}/apps/bash-script/*" ./apps/bash-script

Reviewer: Hmm....let me look at this again

> cp -r "${CNB_SAMPLES_PATH}/apps/bash-script/*" ./apps/bash-script

Reviewer: Natalie's suggestion works.  My  suggestion works on linux and darwin if you omit the `*`.  I'm going to commit Natalie's....this is reason 13893 that I don't like to maintain shell scripts :P

```text
sample-stack-run:jammy apps/bash-script
Starting analyzer...
Parsing inputs...
Ensuring privileges...
Executing command...
Timer: Analyzer started at 2024-09-30T07:38:14Z
Image with name "apps/bash-script" not found
Image with name "cnbs/sample-stack-run:jammy" not found
Timer: Analyzer ran for 41.92µs and ended at 2024-09-30T07:38:14Z
Run image info in analyzed metadata is: 
{"Reference":"","Image":"cnbs/sample-stack-run:jammy","Extend":false,"target":{"os":"linux","arch":"amd64"}}
```

Now if you check the `layers` directory, you should have a `analyzed.toml` file with a few null entries.

#### Detect

In this phase, the `detector` looks for an ordered group of buildpacks that will be used during the `build` phase. The `detector` requires an `order.toml` file being present in the `root` directory, which you could derive from `builder.toml` in the `samples` directory while removing the deprecated `stack` section as follows:

```text
cat "${CNB_SAMPLES_PATH}/builders/jammy/builder.toml" | grep -v -i "stack" | sed 's/\.\.\/\.\./\./' > order.toml

```

> `order.toml` files contain a list of groups with each group containing a list of buildpacks. The `detector` reads `order.toml` and looks for the first group that passes the detection process.

Author: @natalieparellano we have links at the end of the section

##### Set buildpacks layout directory

Before running the `detector`, you need to:

1. Create a `buildpacks` directory in the `root` directory

    ```text
    mkdir -p buildpacks
    ```

2. Then you must populate the `buildpacks` directory with your buildpacks of interest.

    > You have to follow the [directory layout][directory layout] defined in the buildpack spec, where each top-level directory is a `buildpack ID` and each second-level directory is a `buildpack version`.

    We will use [`dasel`](http://github.com/tomwright/dasel/) to help us parse toml files.

>     ```command

Author: I didn't run the command so can't really comment on this. 
@AidanDelaney could you take a look at this?

>     ```command

Reviewer: The magic `for` loop creates a directory structure that looks like the following listing.  In short, it takes our sample buildpacks from the source directory and puts them into the directory structure that `lifecycle` expects:

```
$ tree buildpacks/
buildpacks/
├── samples_dotnet-framework
│   └── 0.0.1
│       └── dotnet-framework
│           ├── bin
│           │   ├── build.bat
│           │   └── detect.bat
│           ├── buildpack.toml
│           └── package.toml
├── samples_hello-extensions
│   └── 0.0.1
│       └── hello-extensions
│           ├── bin
│           │   ├── build
│           │   └── detect
│           └── buildpack.toml
├── samples_hello-moon
│   └── 0.0.1
│       └── hello-moon
│           ├── bin
│           │   ├── build
│           │   └── detect
│           ├── buildpack.toml
│           └── README.md
├── samples_hello-moon-windows
│   └── 0.0.1
│       └── hello-moon-windows
│           ├── bin
│           │   ├── build.bat
│           │   └── detect.bat
│           ├── buildpack.toml
│           └── README.md
├── samples_hello-processes
│   └── 0.0.1
│       └── hello-processes
│           ├── bin
│           │   ├── build
│           │   └── detect
│           ├── buildpack.toml
│           └── README.md
├── samples_hello-universe
│   └── 0.0.1
│       └── hello-universe
│           ├── buildpack.toml
│           ├── package.toml
│           └── README.md
├── samples_hello-universe-windows
│   └── 0.0.1
│       └── hello-universe-windows
│           ├── buildpack.toml
│           ├── package.toml
│           └── README.md
├── samples_hello-world
│   └── 0.0.1
│       └── hello-world
│           ├── bin
│           │   ├── build
│           │   └── detect
│           ├── buildpack.toml
│           ├── package.toml
│           └── README.md
├── samples_hello-world-windows
│   └── 0.0.1
│       └── hello-world-windows
│           ├── bin
│           │   ├── build.bat
│           │   └── detect.bat
│           ├── buildpack.toml
│           ├── package.toml
│           └── README.md
├── samples_java-maven
│   └── 0.0.2
│       └── java-maven
│           ├── bin
│           │   ├── build
│           │   └── detect
│           ├── buildpack.toml
│           └── README.md
├── samples_kotlin-gradle
│   └── 0.0.2
│       └── kotlin-gradle
│           ├── bin
│           │   ├── build
│           │   └── detect
│           ├── buildpack.toml
│           └── README.md
└── samples_ruby-bundler
    └── 0.0.1
        └── ruby-bundler
            ├── bin
            │   ├── build
            │   └── detect
            ├── buildpack.toml
            └── README.md

46 directories, 47 files
```

@natalieparellano Do yuu think that explanation and the directory listing is sufficient?  We can include the example along with the _difficult to understand_ shell loop to generate the structure.

>     ```command

Reviewer 2: How about

Before running this command, the `samples/buildpacks` directory has the following structure:

```
├── hello-world
│   ├── README.md
│   ├── bin
│   │   ├── build
│   │   └── detect
│   ├── buildpack.toml
│   └── package.toml
... more buildpacks ...
```

After, it has the following structure:
```
├── samples_hello-world 
│ └── 0.0.1
│  └── hello-world
│  ├── bin
│  │ ├── build
│  │ └── detect
│  ├── buildpack.toml
│  ├── package.toml
│  └── README.md
... more buildpacks ...
```

```text
Starting detector...
Parsing inputs...
Ensuring privileges...
Executing command...
Timer: Detector started at 2024-10-01T07:00:50Z
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
target distro name/version labels not found, reading /etc/os-release file
Checking for match against descriptor: {linux   []}
======== Results ========
fail: samples/java-maven@0.0.2
======== Results ========
fail: samples/kotlin-gradle@0.0.2
======== Results ========
fail: samples/ruby-bundler@0.0.1
======== Results ========
pass: samples/hello-world@0.0.1
pass: samples/hello-moon@0.0.1
Resolving plan... (try #1)
samples/hello-world 0.0.1
samples/hello-moon  0.0.1
Timer: Detector ran for 26.011769ms and ended at 2024-10-01T07:00:50Z
```

> You can view more details about the [order](https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/detect/#ordertoml), [group](https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/detect/#grouptoml) and [plan](https://buildpacks.io/docs/concepts/components/lifecycle/detect/#plantoml) toml files in the platform documentation.

Reviewer: Can we link to https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/analyze/ for the explanation of what `analyzer` does?

> You can view more details about the [order](https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/detect/#ordertoml), [group](https://buildpacks.io/docs/for-platform-operators/concepts/lifecycle/detect/#grouptoml) and [plan](https://buildpacks.io/docs/concepts/components/lifecycle/detect/#plantoml) toml files in the platform documentation.

Reviewer 2: That works for me - in general, we could add those links for every phase

#### Restore

The `restorer` retrieves cache contents, if it contains any, into the build container. During this phase, the `restorer` looks for layers that could be reused or should be replaced while building the application image.

> First, you need to create the `cache` directory, and then run the `restorer` binary as added below:

Author: updated

```text
mkdir cache
```

```text

> ${CNB_LIFECYCLE_PATH}/restorer -log-level debug -layers="./layers" -group="./layers/group.toml" -cache-dir="./cache" -analyzed="./layers/analyzed.toml"

Reviewer 2: ```suggestion
${CNB_LIFECYCLE_PATH}/restorer -log-level debug -daemon -layers="./layers" -group="./layers/group.toml" -cache-dir="./cache" -analyzed="./layers/analyzed.toml"
```

Reviewer 2: This fixes `ERROR: failed to resolve inputs: run image not found in analyzed metadata`

Timer: Restorer started at 2024-10-01T07:03:47Z
Restoring Layer Metadata
Reading buildpack directory: /tmp/example/layers/samples_hello-world
Reading buildpack directory: /tmp/example/layers/samples_hello-moon
Reading Buildpack Layers directory /tmp/example/layers
Reading buildpack directory: /tmp/example/layers/samples_hello-world
Reading Buildpack Layers directory /tmp/example/layers
Reading buildpack directory: /tmp/example/layers/samples_hello-moon
Timer: Restorer ran for 274.41µs and ended at 2024-10-01T07:03:47Z
```

#### Build

The `builder` transforms application source code into runnable artifacts that can be packaged into a container.

Before running the `builder`, the following steps are required:

1. Create two directories:
   * `platform` directory to store configurations and environment variables
   * `workspace` directory to store application source code and where you build it
  
    ```text
    mkdir -p platform
    mkdir -p workspace
    ```

2. Copy the source code from the `app` directory to the `workspace` directory

    ```text
    cp -r apps/bash-script/* ./workspace
    ```

> 3. Create a `launcher` file with instructions to run your application

Reviewer: Ack.  Can we keep this right now and create an issue for me to update the samples?

> 3. Create a `launcher` file with instructions to run your application

Reviewer 2: Discussed offline, we'll create an issue and add a note here that we're faking the work of the buildpack

       declare -x PWD="/tmp/example/workspace"
       declare -x SHLVL="1"
     layers_dir: /tmp/example/layers/samples_hello-moon
     plan_path: /tmp/samples_hello-moon-3356528850/samples_hello-moon/plan.toml
     plan contents:
---> Done
Processing layers
Updating environment
Reading output files
Updating buildpack processes
Updating process list
Finished running build for buildpack samples/hello-moon@0.0.1
Copying SBOM files
Creating SBOM files for legacy BOM
Listing processes
Timer: Builder ran for 20.200892ms and ended at 2024-10-01T07:07:46Z
```

#### Export

The purpose of the `export` phase is to create a new `OCI` image using a combination of remote layers, local `<layers>/<layer>` layers, and the processed `app` directory.

To export the artifacts built by the `builder`, you first need to specify the path of the `launcher` that your image is going to run:

* For AMD64 architectures
  
  ```text
  export CNB_LINUX_LAUNCHER_PATH=/<your-path>/lifecycle/out/linux-amd64/lifecycle/launcher
  ```

* For ARM64 Architectures

  ```text
  export CNB_LINUX_LAUNCHER_PATH=/<your-path>/lifecycle/out/linux-arm64/lifecycle/launcher
  ```

Now you can run the `exporter`:

```text

> ${CNB_LIFECYCLE_PATH}/exporter --log-level debug -launch-cache "./cache" -daemon -cache-dir "./cache" -analyzed "./layers/analyzed.toml" -group "./layers/group.toml" -layers="./layers" -app "./workspace" -launcher="${CNB_LINUX_LAUNCHER_PATH}" -process-type="shell" apps/bash-script

Reviewer 2: ```suggestion
${CNB_LIFECYCLE_PATH}/exporter --log-level debug -launch-cache "./cache" -daemon -cache-dir "./cache" -analyzed "./layers/analyzed.toml" -group "./layers/group.toml" -layers="./layers" -app "./workspace" -launcher="${CNB_LINUX_LAUNCHER_PATH}" apps/bash-script
```

If we don't hand-create that `./layers/samples_hello-moon/launch.toml` file as I suggest above, we should remove the `-process-type` flag

Reviewer 2: I also see 

```
Warning: Failed to export cache: setting cache metadata: creating metadata file '/Users/na001548/workspace/scratch/build-root/cache/staging/io.buildpacks.lifecycle.cache.metadata': open /Users/na001548/workspace/scratch/build-root/cache/staging/io.buildpacks.lifecycle.cache.metadata: no such file or director
```

And, IDK why (we should investigate cc @AidanDelaney) but the build doesn't fail which is as intended

Author: I updated the command; please let me know if this is correct.

```

You can verify that the image was successfully exported by running the `docker images` command.

### Run the app image

Finally, you can run the exported image as follows:

```text

> docker run -it apps/bash-script ./app.sh

Reviewer 2: ```suggestion
docker run -it apps/bash-script ./app.sh
```

Worth noting, `./app.sh` is necessary if no buildpack specifies a process type, but if we find a way to make the hello world buildpack do it then we can omit it

Author: @natalieparellano not sure I understand your comment or what should be changed?

Reviewer: This is for me to fix.  It has implications wider than this tutorial and I need to check that our test suite passes.

```

```text
OUTPUT PLACEHOLDER
```

## Wrapping up

> At the end of this tutorial, we hope that you now have a better understanding on how you could use `Buildpacks` to create container images. You are now ready to explore this technology further and customize it to fit your application development and deployment needs.

Reviewer 2: ```suggestion
At the end of this tutorial, we hope that you now have a better understanding of what happens during each lifecycle phase, and how you could use `Buildpacks` to create container images. You are now ready to explore this technology further and customize it to fit your application development and deployment needs.
```

* `CNB_PLATFORM_API` or the `Platform API` version, varies depending on the use case. This tutorial uses `v0.14`, which is the latest [Platform API][Platform API] version.
* `CNB_SAMPLES_PATH` represents the path of our local copy of the `samples` directory.
* `CNB_LIFECYCLE_PATH` represents the path of our local compiled `lifecycle` directory.

> Please note that  we only run the commands above on a host machine for the purpose of this tutorial, which is not a common practice for buildpacks.

### Examine lifecycle phases

A single app image build consists of the following phases:

1. [Analysis](#analyze)
2. [Detection](#detect)
3. [Cache Restoration](#restore)
4. [Build](#build)
5. [Export](#export)

> Note that a `platform` executes the phases above either by invoking phase-specific lifecycle binaries in order or by executing `/cnb/lifecycle/creator`.

Let's expand each `lifecycle` phase to explain how the `lifecycle` orchestrates buildpacks:

#### Analyze

The `analyze` phase runs before the `detect` phase in order to validate registry access for all images used during the `build` as early as possible. In this way it provides faster failures for end users.

> Prior to executing `/cnb/lifecycle/analyzer`, you need to create two directories in the `root` directory as follows:

Reviewer 2: ```suggestion
Prior to executing `/cnb/lifecycle/analyzer`, you need to create a parent directory for this tutorial and two directories inside this directory as follows:
```

Author: updated

```text

> mkdir -p apps/bash-script
> mkdir -p layers

Reviewer 2: ```suggestion
mkdir /tmp/tutorial # or your preferred directory
cd /tmp/tutorial
mkdir -p apps/bash-script
mkdir -p layers
```

Author: added

Timer: Restorer started at 2024-10-01T07:03:47Z
Restoring Layer Metadata
Reading buildpack directory: /tmp/example/layers/samples_hello-world
Reading buildpack directory: /tmp/example/layers/samples_hello-moon
Reading Buildpack Layers directory /tmp/example/layers
Reading buildpack directory: /tmp/example/layers/samples_hello-world
Reading Buildpack Layers directory /tmp/example/layers
Reading buildpack directory: /tmp/example/layers/samples_hello-moon
Timer: Restorer ran for 274.41µs and ended at 2024-10-01T07:03:47Z
```

#### Build

The `builder` run buildpacks, that do the actual work of transforming application source code into runnable artifacts.

Before running the `builder`, the following steps are required:

1. Create two directories:
   * `platform` directory to store platform configuration and environment variables
   * `workspace` directory to store the application source code
  
    ```text
    mkdir -p platform
    mkdir -p workspace
    ```

2. Copy the source code from the `app` directory to the `workspace` directory

    ```text
    cp -r apps/bash-script/* ./workspace
    ```

> 3. Create a `launcher` file with instructions to run your application

Reviewer 2: ```suggestion
3. Create a `launcher` file with instructions to run your application; note that in a real buildpacks build, the platform does not create this file! The samples/hello-moon buildpack would create it. In our case, the samples/hello-moon buildpack hasn't been updated with this functionality, so we are faking that behavior.
```

Author: added
